Adding email rate limiting & J5 update

MarkRS-UK opened the pull request:
Hi Jan,
As we discussed (a while ago now) here is my first attempt at updating this to Joomla 5.
I have checked that all the displays work, but not the functionality. I don't (think?) I have the setup to do that.

There is one problem I know of.  If this is used to update an existing 5.0.0 installation the "Send Email" and "Mailing List" links on the control panel don't work, although they work just fine a) from the menu and b) on a clean installation.  I will fix this, but didn't want to delay further.

I have also included the rate limiting (js) code that Mario put in the discussion page, although I haven't checked it.
I have made it user configurable by adding a config tab for it.

I am happy to work on this further but I have to spend more time on my other projects at the moment.
Let me know what you think.

All the best
Mark Stanton

MarkRS-UK commented:
BTW, remembering that you are considering updates for J! 6.0 I have given this version number 5.5.0 although, as a major update, I s'pose it really should be 6.0.0

PhocaCz commented:
Hi, thank you for your work, what exactly does this mean: " the "Send Email" and "Mailing List" links on the control panel don't work"?

MarkRS-UK commented:
The icons on the control panel for those two options produced an error rather than the correct page.  This only happened when I installed it as an update to 5.0

PhocaCz commented:
What type of error?

MarkRS-UK commented:
I get (for Send Email)

```
404 View not found [name, type, prefix]: write, html, Administrator

Call Stack

# | Function | Location
-- | -- | --
1 | () | JROOT/libraries/src/MVC/Controller/BaseController.php:881
2 | Joomla\CMS\MVC\Controller\BaseController->getView() | JROOT/libraries/src/MVC/Controller/BaseController.php:655
3 | Joomla\CMS\MVC\Controller\BaseController->display() | JROOT/administrator/components/com_phocaemail/src/Controller/DisplayController.php:57
4 | Phoca\Component\phocaemail\Administrator\Controller\DisplayController->display() | JROOT/libraries/src/MVC/Controller/BaseController.php:730
5 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT/libraries/src/Dispatcher/ComponentDispatcher.php:143
6 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() | JROOT/libraries/src/Component/ComponentHelper.php:361
7 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT/libraries/src/Application/AdministratorApplication.php:150
8 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT/libraries/src/Application/AdministratorApplication.php:205
9 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT/libraries/src/Application/CMSApplication.php:306
10 | Joomla\CMS\Application\CMSApplication->execute() | JROOT/administrator/includes/app.php:58
11 | require_once() | JROOT/administrator/index.php:32
```

PhocaCz commented:
Hi, then it seems like that the link creation is not correct, but hard to say, I don't know the new methods :idea:

MarkRS-UK commented:
It seems I was wrong about when it happens, which made it easier to find.  Now fixed.
I've updated my forked copy, does that automatically feed through to this PR or do I have to do something else?  It's just a change to the control panel view (in admin/src/View/Cp/HtmlView.php), changing the names of those two views to be called.

PhocaCz commented:
![image](https://github.com/user-attachments/assets/37955322-bb8c-4c3a-bcbc-44794b0869f5)

![image](https://github.com/user-attachments/assets/df530ad1-5d31-4858-a34d-4e9181bce1b4)

![image](https://github.com/user-attachments/assets/e44ff978-685d-4368-8e88-d351a0308621)

![image](https://github.com/user-attachments/assets/edebad91-eb59-42b8-abeb-d38509219b5f)

![image](https://github.com/user-attachments/assets/08152fc2-42bc-40f7-b105-8657bc7fa838)

![image](https://github.com/user-attachments/assets/eafee3f3-0694-467d-82dc-8186820bede0)

![image](https://github.com/user-attachments/assets/411db14a-1fcf-418f-8101-06c91ca446ea)

![image](https://github.com/user-attachments/assets/f138631f-e648-42d1-8f07-5406ae658174)

![image](https://github.com/user-attachments/assets/c5252013-c8dc-44ec-b665-e7e917b38116)

SEF not working

MarkRS-UK commented:
Hi Jan, are you progressing this or do you want me to fix it?
If the latter, could you tell me what you were doing to get this error.

PhocaCz commented:
Hi, I'm still working on it, there are many errors, from typos to missing entries, such as interfaces for routers, etc. I basically have everything resolved, now I will test it in more environments and release the next version. The only thing I couldn't deal with was the Javascript error, so I temporarily commented out the new parameter. That can be resolved later.

This is a valuable lesson for me, because rewriting a Joomla component to a new type of notation means:

- nothing new for the end user - nothing changes for anyone, so no added value
- more chaos and clutter in the file system (e.g. separation of views from tmpl)
- some notations have no alternative, so there remains a mix of old and new
- more meaningless abstraction
- more errors, because many errors arise during rewriting

The result is that we get the same functionality, without added value and with more errors.

So I will prepare new version and will be great if you can then take a look at the Javascript part (the ajax sending emails) so it will not produce errors, so we can add back the limit parameters and it works the way you suggested.

Anyway thank you for the hard work of rewriting, this was really not so easy.

MarkRS-UK commented:
Glad you're on it, and that this was a useful thing for you.
It's not really my area but I'm happy to look at the js stuff when you're ready.  Let me know when.  Incidentally, I didn't write it, it's a suggestion from your forum.
I'm finding with my own components that, although it's more indirect in places, the new systems help me write more robust code, although I don't always accept the help.  I think particularly the namespacing really helps, particularly in an environment like Joomla where we have no idea how much other code (ie other components/modules/plugins) could be around.

PhocaCz commented:
Yes, I agree, I always look at it from the end user's perspective, they usually lose some components that developers don't want to rewrite anymore. From this perspective, I simply can't be excited about the new system.

I'm currently doing the last tests, I would release version 5.1.0 and then I would like you to take a look at the ajax javascript, unfortunately I didn't get it working, so the old one is still there.

MarkRS-UK commented:
Regarding "not losing components", I was trying to write code in a way that could be a useful example across the whole Phoca spectrum.  Not sure if I succeeded or not.

Please could you let me know when it's ready for me to take a look.  I'm not working in a Joomla system at the moment so probably won't see the new release notification.

PhocaCz commented:
BTW, as you can see, namespacing does not solve things by itself, for example modules just don't know the path:

![image](https://github.com/user-attachments/assets/2229f947-0de4-4a2c-8a32-1e4459342c1f)

And still need to solve mysterious problems with new system like doubling output, etc:

![image](https://github.com/user-attachments/assets/fec86f8f-cbe2-4948-88b5-4af28304f00d)

PhocaCz commented:
What is very interesting is that this only happens in Joomla 6 and not in Joomla 5.

![image](https://github.com/user-attachments/assets/2c31ee0e-ed99-4a5b-86bf-1f20ff182db5)

EDIT: I found it, there were duplicities in output files and ignored layout file (the changes must be reverted there)

PhocaCz commented:
https://github.com/PhocaCz/PhocaEmail/releases/download/5.1.0/com_phocaemail_v5.1.0.zip

https://github.com/PhocaCz/PhocaEmailNewsletterModule/releases/download/5.1.0/mod_phocaemail_newsletter_v5.1.0.zip

MarkRS-UK commented:
Ok, I'll get onto it over the next couple of days.

MarkRS-UK commented:
Attached is the fixed sendNewsletter tmpl file.  It's a .txt file because github won't let me upload a php file, just rename it.

This is the only file to change apart from config.xml.  You've still got the addition required in there, just uncomment it.
I tested this with J5.
[default.txt](https://github.com/user-attachments/files/20792660/default.txt)

PhocaCz commented:
Hi, thank you, email rate implemented in version 5.1.1, see:

Phoca Email version 5.1.1 has been released (Phoca Email component and Phoca Email Newsletter module).
https://www.phoca.cz/news/1390-phoca-email-version-5-1-1-released-joomla-5-joomla-6

<img width="1200" height="630" alt="phoca-email-joomla6" src="https://github.com/user-attachments/assets/9ab891db-fddb-4f10-b5c6-955843c880f3" />

PhocaCz commented:
BTW, you might be wondering why the extension rewritten under the new MVC does not work with the gzip setting in Joomla Global Configuration. Access to this page will not be displayed due to bad coding: index.php?option=com_phocaemail&view=newsletter

PhocaCz commented:
OK, found, some hidden characters in Router.php after closing php tag. Will be fixed in next version
